src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -81,9 +83,74 @@ var {
   resetContext,
 } = require('ReactFiberContext');
 
+var onCommitRootInDev = null;
+var callCommitListenerInDev = null;
+
 if (__DEV__) {
   var ReactFiberInstrumentation = require('ReactFiberInstrumentation');
   var ReactDebugCurrentFiber = require('ReactDebugCurrentFiber');
+  var warning = require('warning');
+
+  var devCommitListeners = [];
+  var devMountedRoots = new Set();
```

Comment:
This is unfortunate because it removes a neat GC clean up opportunity (although many subscriptions will hold onto the tree regardless). At least if a whole subtree can be GC:ed, right now we get the clean up automatically (if someone forgets unmountComponentAtNode).

The `__REACT_DEVTOOLS_GLOBAL_HOOK__` is always available regardless if the tool is available or not. That gives us an opportunity to react to its existence. We know that we'll never need to drop any roots because we can just pass them along to the `__REACT_DEVTOOLS_GLOBAL_HOOK__` and it can decide whether we need to keep them around for when the devtools starts.